pkg/hyper/network.go

```diff
@@ -0,0 +1,247 @@
+	Ip string
+}
+
+type CNINetworkPlugin struct {
+	sandboxNs map[string]*sandboxNetNS
+}
+
+func InitCNI() (*CNINetworkPlugin, error) {
+	plugin := &CNINetworkPlugin{}
+
+	plugin.sandboxNs = make(map[string]*sandboxNetNS)
+
+	return plugin, nil
+}
+
+func (plugin *CNINetworkPlugin) SetupPodNetwork(podname string, netdir string) error {
+	sandboxns := &sandboxNetNS{}
+
+	netns, err := ns.NewNS()
+	if err != nil {
+		return err
+	}
+	sandboxns.netns = netns
+
+	if netdir == "" {
+		netdir = DefaultNetDir
+	}
+	sandboxns.netdir = netdir
+
+	plugin.sandboxNs[podname] = sandboxns
+
+	return plugin.sandboxNs[podname].addNetwork()
+}
+
+func (plugin *CNINetworkPlugin) GetPodNetworkStatus(podname string) ([]*InterfaceInfo, error) {
+	return plugin.sandboxNs[podname].getNetworkStatus()
+}
+
+func (plugin *CNINetworkPlugin) TeardownPodNetwork(podname string) error {
+	err := plugin.sandboxNs[podname].delNetwork()
+	if err != nil {
+		return err
+	}
+
+	err = plugin.sandboxNs[podname].netns.Close()
+	if err != nil {
+		return err
+	}
+
+	delete(plugin.sandboxNs, podname)
+
+	return nil
+}
+
+type sandboxNetNS struct {
+	netns  ns.NetNS
+	netdir string
+}
+
+func (sns *sandboxNetNS) addNetwork() error {
+	files, err := libcni.ConfFiles(sns.netdir)
+	switch {
+	case err != nil:
+		return err
+	case len(files) == 0:
+		return fmt.Errorf("No networks found in %s", sns.netdir)
+	}
+
+	sort.Strings(files)
+
+	cninet := &libcni.CNIConfig{
+		Path: strings.Split(DefaultCNIDir, ":"),
+	}
+
+	for _, confFile := range files {
+		netconf, err := libcni.ConfFromFile(confFile)
+		if err != nil {
+			continue
+		}
+
+		rt := &libcni.RuntimeConf{
+			ContainerID: "cni",
+			NetNS:       sns.netns.Path(),
+			IfName:      DefaultInterfaceName,
+		}
+
+		_, err = cninet.AddNetwork(netconf, rt)
+		if err != nil {
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (sns *sandboxNetNS) getNetworkStatus() ([]*InterfaceInfo, error) {
+	var infos []*InterfaceInfo
+	err := sns.netns.Do(func(_ ns.NetNS) error {
+		infos = collectionInterfaceInfo()
+		return nil
+	})
+
+	return infos, err
+}
+
+func (sns *sandboxNetNS) delNetwork() error {
+	files, err := libcni.ConfFiles(sns.netdir)
+	switch {
+	case err != nil:
+		return err
+	case len(files) == 0:
+		return fmt.Errorf("No networks found in %s", sns.netdir)
+	}
+
+	sort.Strings(files)
+
+	cninet := &libcni.CNIConfig{
+		Path: strings.Split(DefaultCNIDir, ":"),
+	}
+
+	for _, confFile := range files {
+		netconf, err := libcni.ConfFromFile(confFile)
+		if err != nil {
+			continue
+		}
+
+		rt := &libcni.RuntimeConf{
+			ContainerID: "cni",
+			NetNS:       sns.netns.Path(),
+			IfName:      DefaultInterfaceName,
+		}
+
+		err = cninet.DelNetwork(netconf, rt)
+		if err != nil {
+			continue
+		}
+	}
+
+	return nil
+}
+
+func collectionInterfaceInfo() []*InterfaceInfo {
+	infos := []*InterfaceInfo{}
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		return infos
+	}
+
+	for _, link := range links {
+		if link.Type() != "veth" {
```

Comment:
In order to support other interface types, it's better to omit only `lo` here, e.g. `interface name == lo`